### PR TITLE
Fix linspace optimization

### PR DESCRIPTION
**Pull request by SimonHeybrock**
- Fix out-of-bounds write when binning by 2 coords
- Add tests, highlighting that current implementation drops events (currently failing tests).

How should we address the latter? We can add special-case handling to make sure we always check upper/lower bounds exactly? Strictly speaking we will still get events in the *wrong* bin in the "interior", so do we also have to address that? We could use the linspace-optimization to provide a "starting guess" for a search (would probably need to consider only 2 or 3 bins). How much performance would we lose?

### CODE REVIEW

**Comment by SimonHeybrock** [comment]
I pushed an update with a potential solution for `histogram`. This appears to cause a 30% performance regression. Does someone have a better idea? Would this be acceptable?

**Comment by SimonHeybrock** [comment]
> This appears to cause a 30% performance regression. Does someone have a better idea? Would this be acceptable?

Actually this might affect mainly the non-threaded case (histogramming a 1d table). With an extra dim or previous binning I see much less difference.

File: lib/core/include/scipp/core/element/histogram.h [hunk 1]
```diff
@@ -61,9 +62,17 @@ static constexpr auto histogram = overloaded{
         const auto [offset, nbin, scale] = core::linear_edge_params(edges);
         for (scipp::index i = 0; i < scipp::size(events); ++i) {
           const auto x = events[i];
-          const double bin = (x - offset) * scale;
-          if (bin >= 0.0 && bin < nbin)
-            iadd(data, static_cast<scipp::index>(bin), weights, i);
+          scipp::index bin = (x - offset) * scale;
```
Comment (nvaytet) [inline lib/core/include/scipp/core/element/histogram.h:65]: Is this now performing the `static_cast` implicitly?

File: tests/hypothesis/bin_test.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# SPDX-License-Identifier: BSD-3-Clause
```
Comment (nvaytet) [inline tests/hypothesis/bin_test.py:1]: So were you only able to cover these edge cases with hypothesis, or can you also add more deterministic tests?

File: lib/core/include/scipp/core/element/histogram.h (hunk 1 above)
Comment (SimonHeybrock) [reply lib/core/include/scipp/core/element/histogram.h:65]: Yes.

File: tests/hypothesis/bin_test.py (hunk 2 above)
Comment (SimonHeybrock) [reply tests/hypothesis/bin_test.py:1]: I think it could be made more deterministic, yes.

Comment (SimonHeybrock) [reply tests/hypothesis/bin_test.py:1]: See update.

File: tests/hypothesis/histogram_test.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+@given(st.floats(**float_args), st.floats(**float_args))
+def test_histogram_linspace_bounds(a, b):
+    x = sc.array(dims=['row'], values=[a, b])
+    table = sc.DataArray(sc.ones(dims=['row'], shape=[2]))
+    table.coords['x'] = x
+    upper = x.max()
+    upper.value = np.nextafter(upper.value, np.inf)
+    edges = sc.linspace('x', x.min(), upper, num=3)
+    hist = sc.histogram(table, bins=edges)
+    assert hist.sum().value == 2
```
Comment (nvaytet) [inline tests/hypothesis/histogram_test.py:26]: Update release notes?

**Comment by jl-wynen** [comment]
Ran `histogram_banchmark` on my workstation and saw no difference between this branch and main.

File: docs/about/release-notes.rst [hunk 4]
```diff
@@ -21,6 +21,9 @@ Bugfixes
 ~~~~~~~~
 
 * Fix bug where coordinates in were dropped by :func:`scipp.transform_coords` with ``keep_*=False`` even through they were specified as ``targets`` `#2642 <https://github.com/scipp/scipp/pull/2642>`_.
+* Fix segmentation fault in :func:`scipp.bin` when binning in 2 or more dimensions withs certain coordinate values `#2644 <https://github.com/scipp/scipp/pull/2644>`_.
```
Comment (nvaytet) [inline docs/about/release-notes.rst:24]:
```suggestion
* Fix segmentation fault in :func:`scipp.bin` when binning in 2 or more dimensions with certain coordinate values `#2644 <https://github.com/scipp/scipp/pull/2644>`_.
```

File: docs/about/release-notes.rst [hunk 5]
```diff
@@ -21,6 +21,9 @@ Bugfixes
...
+* Fix issue with events close to upper or lower bin bounds getting dropped by :func:`scipp.bin` and :func:`scipp.histogram` when binning with edges that form a "linspace" `#2644 <https://github.com/scipp/scipp/pull/2644>`_.
+* Fix minor issue with events close bin bounds getting assign to wrong bin by :func:`scipp.bin` and :func:`scipp.histogram` when binning with edges that form a "linspace" `#2644 <https://github.com/scipp/scipp/pull/2644>`_.
```
Comment (nvaytet) [inline docs/about/release-notes.rst:26]:
```suggestion
* Fix minor issue with events close to bin bounds getting assigned to the wrong bin by :func:`scipp.bin` and :func:`scipp.histogram` when binning with edges that form a "linspace" `#2644 <https://github.com/scipp/scipp/pull/2644>`_.
```

**Review by jl-wynen — APPROVED** [review]
